File: 30_oct/problem_01_balanced_brackets/solution.py
from typing import TextIO
import sys

class Solution:
    def areBracketsBalanced(self, s):
        stack = []
        bracket_map = {")": "(", "}": "{", "]": "["}
        
        for char in s:
            if char in bracket_map:
                top_element = stack.pop() if stack else '#'
                if bracket_map[char] != top_element:
                    return False
            else:
                stack.append(char)
                
        return not stack

def solve(infile: TextIO = sys.stdin, outfile: TextIO = sys.stdout) -> None:
    s = infile.readline().strip()
    
    solution = Solution()
    result = solution.areBracketsBalanced(s)
    
    outfile.write("true\n" if result else "false\n")

if __name__ == "__main__":
    solve()

File: 30_oct/problem_01_balanced_brackets/solution.cpp
using namespace std;
#include <iostream>
#include <string>
#include <stack>
#include <map>

class Solution {
public:
    bool areBracketsBalanced(const string& s) {
        stack<char> st;
        map<char, char> bracket_map;
        bracket_map[')'] = '(';
        bracket_map['}'] = '{';
        bracket_map[']'] = '[';

        for (int i = 0; i < s.length(); ++i) {
            char c = s[i];
            if (bracket_map.find(c) == bracket_map.end()) {
                st.push(c);
            } else {
                if (st.empty() || st.top() != bracket_map[c]) {
                    return false;
                }
                st.pop();
            }
        }
        return st.empty();
    }
};

int main() {
    string s;
    getline(cin, s);
    
    Solution sol;
    if (sol.areBracketsBalanced(s)) {
        cout << "true" << endl;
    } else {
        cout << "false" << endl;
    }
    
    return 0;
}
